internal/cmd/agent/controllers/bundledeployment/controller_test.go

```diff
@@ -0,0 +1,119 @@
+			expectedNs: corev1.Namespace{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels:      map[string]string{"name": "test", "optLabel1": "optValue1", "optLabel2": "optValue2"},
+					Annotations: map[string]string{"optAnn1": "optValue1"},
+				},
+			},
+			expectedErr: nil,
+		},
+
+		"NamespaceLabels and NamespaceAnnotations removes entries that are not in the options, except the name label": {
+			bd: &fleet.BundleDeployment{Spec: fleet.BundleDeploymentSpec{
+				Options: fleet.BundleDeploymentOptions{
+					NamespaceLabels:      &map[string]string{"optLabel": "optValue"},
+					NamespaceAnnotations: &map[string]string{},
+				},
+			}},
+			ns: corev1.Namespace{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels:      map[string]string{"nsLabel": "nsValue", "name": "test"},
+					Annotations: map[string]string{"nsAnn": "nsValue"},
+				},
+			},
+			release: "namespace/foo/bar",
+			expectedNs: corev1.Namespace{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels:      map[string]string{"optLabel": "optValue", "name": "test"},
+					Annotations: map[string]string{},
+				},
+			},
+			expectedErr: nil,
+		},
+
+		"NamespaceLabels and NamespaceAnnotations updates existing values": {
+			bd: &fleet.BundleDeployment{Spec: fleet.BundleDeploymentSpec{
+				Options: fleet.BundleDeploymentOptions{
+					NamespaceLabels:      &map[string]string{"bdLabel": "labelUpdated"},
+					NamespaceAnnotations: &map[string]string{"bdAnn": "annUpdated"},
+				},
+			}},
+			ns: corev1.Namespace{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels:      map[string]string{"bdLabel": "nsValue"},
+					Annotations: map[string]string{"bdAnn": "nsValue"},
+				},
+			},
+			release: "namespace/foo/bar",
+			expectedNs: corev1.Namespace{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels:      map[string]string{"bdLabel": "labelUpdated"},
+					Annotations: map[string]string{"bdAnn": "annUpdated"},
+				},
+			},
+			expectedErr: nil,
+		},
+	}
+
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+			ctrl := gomock.NewController(t)
+			defer ctrl.Finish()
+
+			mockDynamic := mocks.NewMockInterface(ctrl)
+			mockNamespaceableResourceInterface := mocks.NewMockNamespaceableResourceInterface(ctrl)
+			u, _ := runtime.DefaultUnstructuredConverter.ToUnstructured(&test.ns)
+			mockDynamic.EXPECT().Resource(gomock.Any()).Return(mockNamespaceableResourceInterface).Times(2)
+			mockNamespaceableResourceInterface.EXPECT().List(gomock.Any(), metav1.ListOptions{
+				LabelSelector: "name=namespace",
+			}).Return(&unstructured.UnstructuredList{
+				Items: []unstructured.Unstructured{{Object: u}},
+			}, nil).Times(1)
+			uns, _ := runtime.DefaultUnstructuredConverter.ToUnstructured(&test.expectedNs)
+			mockNamespaceableResourceInterface.EXPECT().Update(gomock.Any(), &unstructured.Unstructured{Object: uns}, gomock.Any()).Times(1)
+
+			h := handler{
+				dynamic: mockDynamic,
+			}
+			err := h.setNamespaceLabelsAndAnnotations(test.bd, test.release)
```

Comment:
Maybe something for a wider discussion: I'm wondering about testing unexported methods, not least because I like having tests isolated in `$package_test` packages, but also because if at some point we omit or comment out the call to `setNamespaceLabelsAndAnnotations` in the controller, this test driver will still pass although Fleet will not behave as we want it to.
Should we add a higher-level test case here, of course without repeating all combinations, but just to make sure that labels and annotations are altered when deploying a namespace from a bundle?

Reply:
I wish we could. However the current code doesn't allow that, a refactor would need to be done in other to inject mocks for all dependencies.

Reply:
I see; and an end-to-end test case for that could be a bit overkill.
It could be nice to follow the Golang proverb "accept interfaces, return structs" to ease mocking, which of course is somewhat beyond the scope of this PR :)

Reply:
Indeed, would be great if we could use `_test` for tests and do away with the "java-style" interfaces. But a lot of the existing code is that way and wrangler complicates it further. While we're at it, I'd like a proper `context.Context` in the networked methods. I think wrangler puts one in a struct, but it's not passed into the handlers.